File: tests/fixtures/v2-redirect-flow/positive/redirect_flow.py
def direct(request):
    return redirect(request.args["direct"])


def propagated(request):
    requested = request.args["propagated"]
    alias = requested
    return redirect(alias)


def parsed(request):
    destination = urllib.parse.urlsplit(request.args["parsed"])
    return redirect(destination.geturl())


def valid_local(destination, allowed_hosts):
    return url_has_allowed_host_and_scheme(destination, allowed_hosts=allowed_hosts)

File: tests/fixtures/v2-redirect-flow/positive/RedirectFlow.java
class RedirectFlow {
    void direct(HttpServletRequest request, HttpServletResponse response) throws Exception {
        response.sendRedirect(request.getParameter("direct"));
    }

    void propagated(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String requested = request.getParameter("propagated");
        String alias = requested;
        response.sendRedirect(alias);
    }

    void parsed(HttpServletRequest request, HttpServletResponse response) throws Exception {
        URI destination = URI.create(request.getParameter("parsed"));
        response.sendRedirect(destination.toString());
    }

    boolean validLocal(String destination) {
        return !URI.create(destination).isAbsolute();
    }
}
